crates/typst/src/foundations/calc.rs

```diff
@@ -915,45 +1053,116 @@ pub enum Num {
 }
 
 impl Num {
-    fn apply2(
+    fn float(self) -> f64 {
+        match self {
+            Self::Int(v) => v as f64,
+            Self::Float(v) => v,
+        }
+    }
+}
+
+cast! {
+    Num,
+    self => match self {
+        Self::Int(v) => v.into_value(),
+        Self::Float(v) => v.into_value(),
+    },
+    v: i64 => Self::Int(v),
+    v: f64 => Self::Float(v),
+}
+
+/// A value which can be passed to functions that work with decimals, integers or floats.
+#[derive(Debug, Copy, Clone)]
+pub enum DecNum {
+    Decimal(Decimal),
+    Int(i64),
+    Float(f64),
+}
+
+impl DecNum {
+    /// Checks if this number is equivalent to zero.
+    fn is_zero(self) -> bool {
+        match self {
+            Self::Decimal(d) => d.is_zero(),
+            Self::Int(i) => i == 0,
+            Self::Float(f) => f == 0.0,
+        }
+    }
+
+    /// If this `DecNum` holds an integer or float, returns a float.
+    /// Otherwise, returns `None`.
+    fn float(self) -> Option<f64> {
+        match self {
+            Self::Decimal(_) => None,
+            Self::Int(i) => Some(i as f64),
+            Self::Float(f) => Some(f),
+        }
+    }
+
+    /// If this `DecNum` holds an integer or decimal, returns a decimal.
+    /// Otherwise, returns `None`.
+    fn decimal(self) -> Option<Decimal> {
+        match self {
+            Self::Decimal(d) => Some(d),
+            Self::Int(i) => Some(Decimal::from(i)),
+            Self::Float(_) => None,
+        }
+    }
+
+    /// Tries to apply a function to two decimal or numeric arguments.
+    ///
+    /// Fails with `None` if one is a float and the other is a decimal.
+    fn apply2<T>(
```

Comment:
This function's signature is technically inconsistent with `apply3` below, as it allows for arbitrary return values (`T`). I wonder if we should give it a different name, but I couldn't think of anything better so far.

Reply:
Hmm, yeah. It's ok for now.